### PR TITLE
Add filtering and version extraction pre-validation

### PR DESCRIPTION
**Pull request by mcharriere**
Couldflared added the `-amd64` suffix to all the new images starting from 2022.8.1.

This PR adds support for filtering and version extraction before validating SemVer.
It also adds the proper filter for cloudflared

Towards https://github.com/giantswarm/giantswarm/issues/20258

### CODE REVIEW

File: pkg/retagger/job_pattern.go [hunk 1]
```diff
@@ -28,7 +30,7 @@ type PatternJob struct {
 
 // Compile expands a PatternJob into one or multiple SingleJobs using the given Retagger instance.
 func (job *PatternJob) Compile(r *Retagger) ([]SingleJob, error) {
-	r.logger.Log("level", "debug", "message", fmt.Sprintf("compiling jobs for image %v / %v using pattern %v, with options %#v", job.Source.RepoPath, job.Source.Image, job.SourcePattern, job.Options))
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("compiling jobs for image %v / %v using pattern %v with filter %s, with options %#v", job.Source.RepoPath, job.Source.Image, job.SourcePattern, job.SourceFilter, job.Options))
```
Comment (uvegla) [inline pkg/retagger/job_pattern.go:33]:
```suggestion
	r.logger.Log("level", "debug", "message", fmt.Sprintf("compiling jobs for image %v / %v using pattern %v, with filter %s, with options %#v", job.Source.RepoPath, job.Source.Image, job.SourcePattern, job.SourceFilter, job.Options))
```

File: pkg/retagger/job_pattern.go [hunk 2]
```diff
@@ -123,11 +125,29 @@ func (job *PatternJob) getExternalTagMatches(r *dockerRegistry.Registry, image s
...
+	}
+
 	// Find tags matching our configured pattern.
 	var matches []string
 	for _, t := range externalRegistryTags {
 		job.logger.Log("level", "debug", "message", fmt.Sprintf("Checking external tag: %s ", t))
-		v, err := semver.NewVersion(t)
+
+		ts := t
+		if filter != "" {
```
Comment (uvegla) [inline pkg/retagger/job_pattern.go:139]: Could you please write test for this new piece of functionality?

File: README.md [hunk 3]
```diff
@@ -80,10 +80,12 @@ What the attributes mean:
 - `tags[].customImages[].dockerfileOptions[]`: The list of Dockerfile options, used to override base image
 - `patterns[]`: List of patterns. New tags matching one of these patterns will be automatically retagged.
 - `patterns[].pattern`: Valid semver condition to match tags.
+- `patterns[].filter`: Regex to filter tags before validating semver. If empty, filtering is not done. 
 - `patterns[].customImages[]`: Custom images as explained above.
 
 An image may define both `tags` and `patterns`.
```
Comment (uvegla) [inline README.md:86]:
```suggestion
An image may define both `tags`, `patterns` and `filters`.
```

**Review by uvegla — CHANGES_REQUESTED** [review]
Please add a few unit tests for the new piece of functionality around patterns.

File: README.md (hunk 3 above)
Comment (mcharriere) [reply README.md:86]: tags and patterns are "top level" attributes of image. filter is inside patterns. I don't think it is correct add filters here.

**Comment by mcharriere** [comment]
To sum up, this PR includes:

- New CI job running unitTests before building
- Increase wait time in e2e test (between the push of the images and the check, as Quay takes up to a minute to show them)
- Add filtering and extraction based on a regex
- Add unit tests for find matches and filtering
- Add cloudflared filtering regex and update retagging to newer version

**Review by uvegla — APPROVED** [review]
Thank you! 🙇🏻

**Pull request merged**